### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by lincoln1883**
For this milestone I was able to complete the following tasks:
- Create a class Classroom with the following:
  - [ ] @label instance variable, should be initialized in the constructor.
  - [ ] Setter and getter for @label (remember about attr_accessor).
- Create the has-many/belongs-to relationship between Classroom and Student. The following should be implemented:
  - [ ] Create the has-many side (a classroom has many students).
  - [ ] Create the belongs-to side (a student belongs to a classroom).
  - [ ] Make sure that when adding a student to a classroom it also sets the classroom for the student.
  - [ ] Make sure that when setting the classroom for a student it also adds it to the classroom's students.
- Create a class Book with the following
  - [ ] @title and @author instance variables, should be initialized in the constructor.
  - [ ] Setters and getters for instance variables (remember about attr_accessor).
- Create a class Rental with the following:  
  - [ ] @date instance variable, should be initialized in the constructor.
  - [ ] Setter and getter for @date (remember about attr_accessor).
- Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental.
    - [ ] Create the has-many side of Book and Rental (a book has many rentals).
    - [ ] Create the belongs-to side of Rental and Book (a rental belongs to a book).
    - [ ] Create the has-many side of Person and Rental (a person has many rentals).
    - [ ] Create the belongs-to side of Rental and Person (a rental belongs to a person).
    - [ ] Modify the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: classroom.rb
Comment (brytebee) [inline classroom.rb:1]: https://github.com/lincoln1883/Oop_school_library/blob/bcced23281ca7a905029f7a5c95bf03ebc50d217/student.rb#L1-L15

Kindly have a method that serves as the  ['belongs-to' side (a student belongs to a classroom). ](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships)

As in the example from that article, when a student is added to a classroom, the student is `push`ed into the classroom. Ensure you check if the student exist within the classrom.

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class Book
```
Comment (brytebee) [inline book.rb:2]: The Book class shares a many-to-many relationship with the Person class using a Rental class as the join.

Just like the Person class, the book class should have the following:
 - [rental attr](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#belongs-to-relationships)
 - initialized with a rental collection ie. Array eg. `@rentals = []`
 - [add_rental method which takes the person, date and book(self) params and creates a new rental.](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship)

**Review by brytebee — CHANGES_REQUESTED** [review]
Hi, @lincoln1883 👋,

##

### Highlights💡
- Good commit message.✔️ 
- Good PR Summary.✔️
- Great Ruby syntax.✔️
- Good understanding of associations.✔️

##

Good job so far coming so far with the progress you've made! 💪💪
 
However, there are some issues that you still need to work on to go to the next project but you are almost there!



# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Please, remember to tag me in your question so I can receive the notification.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @brytebee in your question so I can receive the notification.
### _**Use @brytebee on slack for faster responses.**_


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by brytebee** [comment]
@lincoln1883 Kindly add a date to the add rental method. i just updated it.

File: student.rb [hunk 2]
```diff
@@ -5,11 +5,16 @@ class Student < Person
   def initialize(age, name, classroom, parent_permission)
     super(age, name)
     @id = id
-    @classroom = classroom
+    @classroom = Classroom.new(classroom)
```
Comment (Ismailco) [inline student.rb]: - [ ] Good job implementing the `add_to_classroom` method, you're pushing the classroom in the method therefore this line of code is not necessary kindly consider removing the `Classroom.new(classroom)` and replace it with an empty array `[]`

```suggestion
    @classroom = []
```

File: book.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+require_relative 'rental'
+class Book
+  attr_accessor :author, :title, :rental
+
+  def initialize(title, author, rental)
+    @title = title
+    @author = author
+    @rental = rental
```
Comment (Ismailco) [inline book.rb]: - [ ] Kindly make sure the rentals are an empty array instead as the previous reviewer suggested

```suggestion
    @rental = []
```

File: book.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  end
```
Comment (Ismailco) [inline book.rb:9]: - [ ] As the previous reviewer mentioned kindly consider implementing a method `add_rentals` that provides the `many_to_many` relationship, same way you did in the `Person` Class

e.g:
```ruby
def add_rental(person, date)
  Rental.new(date, self, person)
end
```

**Review by Ismailco — CHANGES_REQUESTED** [review]
### Hi @lincoln1883, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights

✔️ No linter errors
✔️ Used Gitflow
✔️ Work documented professionally

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by algerina — APPROVED** [review]
Hi @lincoln1883 ,

Great job so far!  

May I highlight the following: 
- [x] Your  project is setup correctly👏  ✔️ 
- [x] Your implemented all the requirements of the previous TSE✔️ 

## STATUS: APPROVED :muscle::clap:

Congratulations! :tada:

Your project is complete!  It's time to merge it :shipit:


 
 ## Optional suggestions

 - Please check the comment under the review.

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._
Cheers and Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread and tag me @algerina if something is not 100% clear.

#### _**Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.**_
------------------------------------------------------------------------------------------------------------------------------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using this [form.](https://airtable.com/shrQAqnBwek5a0O0s)
___
_If anything seems to be unclear to you, reach out and ask anything by commenting on this Pull Request and mention me as [Algerina](https://github.com/algerina), I will receive a notification and will respond._

[GitHub](https://github.com/algerina) | [LinkedIn](https://linkedin.com/in/amel-khiri) | [Twitter](https://twitter.com/AmelQahwadji)

**Pull request merged**